staging/src/k8s.io/apiserver/pkg/storage/cacher/caching_object.go

```diff
@@ -161,6 +166,25 @@ func (o *cachingObject) GetObject() runtime.Object {
 	return o.object.DeepCopyObject().(metaRuntimeInterface)
 }
 
+// GetCachedEvent implements runtime.CacheableObject interface.
+// It returns a watch event containing the already encoded object.
+func (o *cachingObject) GetCachedEvent(id runtime.Identifier, eventProducer func() (runtime.Object, error)) (runtime.Object, error) {
```

Comment:
This code makes some assumption about lifecycle of cachingObject, e.g. it is never reused between watch events.

It actually seems to be true today, as IIUC cachingObject is always created for very specific watchEvent: https://github.com/kubernetes/kubernetes/blob/fa294f3e4e0006219d8c5ff39a0d3bc0e5ec6496/staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go#L847-L876

I think we need either to remove this assumption (and e.g. use watchType as a key in cachedEvents) or somehow ensure that it will be always true (e.g. adding comment somewhere, preferably some unit test checking this, but this may be hard to do).